broker/src/broker.rs

```diff
@@ -0,0 +1,515 @@
+            }
+        }
+
+        Ok(Response::new(Empty {}))
+    }
+
+    type SubscribeSafekeeperInfoStream =
+        Pin<Box<dyn Stream<Item = Result<SafekeeperTimelineInfo, Status>> + Send + 'static>>;
+
+    async fn subscribe_safekeeper_info(
+        &self,
+        request: Request<SubscribeSafekeeperInfoRequest>,
+    ) -> Result<Response<Self::SubscribeSafekeeperInfoStream>, Status> {
+        let proto_key = request.into_inner().subscription_key.ok_or(Status::new(
+            Code::InvalidArgument,
+            "missing subscription key",
+        ))?;
+        let sub_key = SubscriptionKey::from_proto_subscription_key(proto_key)?;
+        let mut subscriber = self.registry.register_subscriber(sub_key).await;
+
+        // transform rx into stream with item = Result, as method result demands
+        let output = async_stream::try_stream! {
+            while let Some(info) = subscriber.sub_rx.recv().await {
+                    yield info
+                }
+
+            // internal generator
+            // let _ = subscriber.sub_rx.try_recv().ok();
+            // let mut counter = 0;
+            // loop {
+            // let info = SafekeeperTimelineInfo {
+            //     safekeeper_id: 1,
+            //     tenant_timeline_id: Some(ProtoTenantTimelineId {
+            //         tenant_id: vec![0xFF; 16],
+            //         timeline_id: vec![0xFF; 16],
+            //         // timeline_id: tli_from_u64(counter),
+            //     }),
+            //     last_log_term: 0,
+            //     flush_lsn: counter,
+            //     commit_lsn: 2,
+            //     backup_lsn: 3,
+            //     remote_consistent_lsn: 4,
+            //     peer_horizon_lsn: 5,
+            //     safekeeper_connstr: "zenith-1-sk-1.local:7676".to_owned(),
+            // };
+            // counter += 1;
+            // yield info;
+            // }
+        };
+
+        Ok(Response::new(
+            Box::pin(output) as Self::SubscribeSafekeeperInfoStream
+        ))
+    }
+}
+
+#[derive(Parser, Debug)]
+#[command(version, about = "Broker for neon storage nodes communication", long_about = None)]
+struct Args {
+    /// Endpoint to listen on.
+    #[arg(short, long, default_value = DEFAULT_LISTEN_ADDR)]
+    listen_addr: SocketAddr,
+}
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // console_subscriber::init();
+
+    let args = Args::parse();
+
+    let registry = Registry {
+        shared_state: Arc::new(Mutex::new(SharedState {
+            pub_txs: HashMap::new(),
+            next_pub_id: 0,
+            subs_to_all: HashMap::new(),
+            subs_to_timelines: HashMap::new(),
+            next_sub_id: 0,
+        })),
+    };
+    let neon_broker_service = NeonBrokerService {
+        registry: registry.clone(),
+    };
+
+    tokio::spawn(async move { registry.report().await });
+
+    Server::builder()
+        .http2_keepalive_interval(Some(Duration::from_millis(5000)))
```

Comment:
I propose to put both report and this timeouts/intervals and the queue size into `Args`.